src/ripple/app/tx/impl/XChainBridge.cpp

```diff
@@ -0,0 +1,1990 @@
+   them collude, they can steal the funds put into trust.
+
+   This also requires a way to prove that assets were put into the box - either
+   a locking-chain-token on the locking chain or returning an
+   issuing-chain-token on the issuing chain. A set of servers called "witness
+   servers" fill this role. These servers watch the ledger for these
+   transactions, and attest that the given events happened on the different
+   chains by signing messages with the event information.
+
+   There needs to be a way to prevent the attestations from the witness
+   servers from being used more than once. "Claim ids" fill this role. A claim
+   id must be acquired on the destination chain before the asset is "put into
+   the box" on the source chain. This claim id has a unique id, and once it is
+   destroyed it can never exist again (it's a simple counter). The attestations
+   reference this claim id, and are accumulated on the claim id. Once a quorum
+   is reached, funds can move. Once the funds move, the claim id is destroyed.
+
+   Finally, a claim id requires that the sender has an account on the
+   destination chain. For some chains, this can be a problem - especially if
+   the wrapped asset represents XRP, and XRP is needed to create an account.
+   There's a bootstrap problem. To address this, there is a special transaction
+   used to create accounts. This transaction does not require a claim id.
+
+   See the document "docs/bridge/spec.md" for a full description of how
+   bridges and their transactions work.
+*/
+
+namespace {
+
+enum class TransferHelperCanCreateDst { no, yes };
+
+/** Transfer funds from the src account to the dst account
+
+    @param psb The payment sandbox.
+    @param src The source of funds.
+    @param dst The destination for funds.
+    @param dstTag Integer destination tag. Used to check if funds should be
+           transferred to an account with a `RequireDstTag` flag set.
+    @param claimOwner Owner of the claim ledger object.
+    @param amt Amount to transfer from the src account to the dst account.
+    @param canCreate Flag to determine if accounts may be created using this
+   transfer.
+    @param j Log
+
+    @return tesSUCCESS if payment succeeds, otherwise the error code for the
+            failure reason.
+ */
+
+TER
+transferHelper(
+    PaymentSandbox& psb,
+    AccountID const& src,
+    AccountID const& dst,
+    std::optional<std::uint32_t> const& dstTag,
+    std::optional<AccountID> const& claimOwner,
+    STAmount const& amt,
+    TransferHelperCanCreateDst canCreate,
+    beast::Journal j)
+{
+    if (dst == src)
+        return tesSUCCESS;
+
+    auto const dstK = keylet::account(dst);
+    if (auto sleDst = psb.read(dstK))
+    {
+        // Check dst tag and deposit auth
+
+        if ((sleDst->getFlags() & lsfRequireDestTag) && !dstTag)
+            return tecDST_TAG_NEEDED;
+
+        if ((dst != claimOwner) && (sleDst->getFlags() & lsfDepositAuth) &&
```

Comment:
I just stumbled on this by accident.  I think the...
```
(dst != claimOwner)
```
part of the `if` is not correct.  When an account creates a claim they specify very little.  They do specify the

- `OtherChainSource` and
- `SignatureReward`.

But they specifically do _not_ specify...

- The _destination_ of the transfer.

In short, just because an account creates a `ClaimID` does not mean that account is expecting to receive any money from the `Door` account.

So, if the `claimOwner` has the `DepositAuth` flag set on their account, I don't think it's fair to assume that it has authorized transfers from the Door account.  I think `claimOwner` would need to do that explicitly with a `DepositPreauth` of the Door account.

Thoughts?

Reply:
Good catch! Just some history - we do want to bypass the `DepositAuth` check in the `Claim` transaction - because that's the claim owner sending funds to themselves. This probably slipped through when we allowed attestations to move funds themselves, instead of always requiring claim. At any rate, I'll make a patch so this is only used during a `claim` transaction.

Reply:
Thanks for looking.  Yeah, if I just remove that part of the condition in the `if` then other things break.  So there are situations where that check needs to be made.  I didn't see a way to fix that given the arguments that `TransferHelper()` currently receives, but I probably missed something.  Or you may need to pass in an additional argument.

Let me know once you have this fixed the way you want it.  I'll work on other code reviews until that's ready.  Thanks.

Reply:
Fixed in cdad3be4f0 [fold] Only allow deposit auth bypass on claim transaction